pennylane/_qutrit_device.py

```diff
@@ -0,0 +1,390 @@
+                    )
+                # Check if the state is accessible and decide to return the state or the density
+                # matrix.
+                results.append(self.access_state(wires=obs.wires))
+
+            elif obs.return_type is not None:
+                raise qml.QuantumFunctionError(
+                    f"Unsupported return type specified for observable {obs.name}"
+                )
+
+            # TODO: Add support for MutualInfo and VnEntropy return types. Currently, qml.math is hard coded
+            # to calculate these quantities for qubit states, so it needs to be updated before these return types
+            # can be supported for qutrits.
+
+        return results
+
+    def generate_samples(self):
+        r"""Returns the computational basis samples generated for all wires.
+
+        Note that PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
+        :math:`q_0` is the most significant trit.
+
+        .. warning::
+
+            This method should be overwritten on devices that
+            generate their own computational basis samples, with the resulting
+            computational basis samples stored as ``self._samples``.
+
+        Returns:
+             array[complex]: array of samples in the shape ``(dev.shots, dev.num_wires)``
+        """
+        number_of_states = 3**self.num_wires
+
+        rotated_prob = self.analytic_probability()
+
+        samples = self.sample_basis_states(number_of_states, rotated_prob)
+        return QutritDevice.states_to_ternary(samples, self.num_wires)
+
+    @staticmethod
+    def generate_basis_states(num_wires, dtype=np.uint32):
+        """
+        Generates basis states in ternary representation according to the number
+        of wires specified.
+
+        Args:
+            num_wires (int): the number wires
+            dtype=np.uint32 (type): the data type of the arrays to use
+
+        Returns:
+            array[int]: the sampled basis states
+        """
+        # A slower, but less memory intensive method
+        basis_states_generator = itertools.product((0, 1, 2), repeat=num_wires)
+        return np.fromiter(itertools.chain(*basis_states_generator), dtype=dtype).reshape(
+            -1, num_wires
+        )
+
+    @staticmethod
+    def states_to_ternary(samples, num_wires, dtype=np.int64):
+        """Convert basis states from base 10 to ternary representation.
+
+        This is an auxiliary method to the generate_samples method.
+
+        Args:
+            samples (array[int]): samples of basis states in base 10 representation
+            num_wires (int): the number of qutrits
+            dtype (type): Type of the internal integer array to be used. Can be
+                important to specify for large systems for memory allocation
+                purposes.
+
+        Returns:
+            array[int]: basis states in ternary representation
+        """
+        ternary_arr = []
```

Comment:
I think the implementation here could be simplified by leveraging [`np.base_repr`](https://numpy.org/doc/stable/reference/generated/numpy.base_repr.html)? For example,
```
>>> np.base_repr(11, 3)
'102'
```

Reply:
I agree that this would significantly simplify the implementation, but it would no longer be consistent with `QubitDevice.states_to_binary()`, which returns a nested list of 0s and 1s.